### PR TITLE
Add fragment backoffice title and improve validation

### PR DESCRIPTION
**Pull request by qsomazzi**
## Changelog

```markdown
### Added
- Fragment "Backoffice title"

### Changed
- FragmentService `buildEditForm` now call `buildForm` to add backoffice title to each fragments
- FragmentService validate signature
```
## Subject

The goal of this PR is to handle new backoffice title for each fragments. This improve the readability of the admin, instead of seeing only the fragment type, you will see a label related to your content. Also, the validate signature was wrong, because it's called from the article admin.

Before
------
capture d ecran 2016-12-01 a 10 46 49

After 
-----
capture d ecran 2016-12-01 a 10 46 07

### CODE REVIEW

File: Resources/public/js/fragmentList.jquery.js [hunk 1]
```diff
@@ -134,21 +154,26 @@
             e.preventDefault();
             e.stopPropagation();
 
+            var sure = confirm("The fragment will be deleted after update. Are you sure?");
+
+            if (!sure) {
```
Comment (greg0ire) [inline Resources/public/js/fragmentList.jquery.js]: This variable is used only once. Why not just write

```
if (!confirm("The fragment will be deleted after update. Are you sure?")) {
{
}
```

File: Resources/public/js/fragmentList.jquery.js [hunk 2]
```diff
@@ -134,21 +154,26 @@
             e.preventDefault();
             e.stopPropagation();
 
+            var sure = confirm("The fragment will be deleted after update. Are you sure?");
```
Comment (greg0ire) [inline Resources/public/js/fragmentList.jquery.js]: No i18n?

File: Resources/public/js/fragmentList.jquery.js [hunk 3]
```diff
@@ -90,10 +91,29 @@
...
+                    if (!isNaN(fragmentId) && fragmentId > maxIndex) {
+                        maxIndex = parseInt(fragmentId);
+                    }
+                });
+
+                this._elementsIndex = maxIndex;
+            }
+
+            this._elementsIndex++;
+            console.log('[Sonata.Admin] [fragments|add] current index is ' + this._elementsIndex);
```
Comment (rande) [inline Resources/public/js/fragmentList.jquery.js]: ``Admin.log`` and not ``console.log``

**Comment by qsomazzi** [comment]
Hello @sonata-project/contributors, we really need this merge by the morning, can we merge it please ?

**Comment by qsomazzi** [comment]
One required check can't be launched but i don't know why

**Comment by greg0ire** [comment]
I'll talk to @Soullivaneuh about that

**Comment by qsomazzi** [comment]
thanks @greg0ire

**Comment by greg0ire** [comment]
BTW, shouldn't this be split into 2 commits? It looks like you are doing 2 separate things here.

**Comment by greg0ire** [comment]
@Soullivaneuh unstuck StyleCI, but you have to force push to trigger it.

File: FragmentService/AbstractFragmentService.php [hunk 4]
```diff
@@ -109,8 +120,13 @@ public function configureOptions(OptionsResolver $resolver)
     /**
      * {@inheritdoc}
      */
-    public function validate(FragmentInterface $fragment, ExecutionContextInterface $context)
+    public function validate(ErrorElement $errorElement, $object)
     {
+        if (empty($object->getBackofficeTitle())) {
```
Comment (greg0ire) [inline FragmentService/AbstractFragmentService.php:124]: Isn't there a more Symfony-ish way of doing this, by using the `NotBlank` constraint?

Comment (greg0ire) [reply FragmentService/AbstractFragmentService.php:124]: Could save you some translations…

Comment (qsomazzi) [reply FragmentService/AbstractFragmentService.php:124]: the whole validation part is still a work in progress, for the moment we can't use symfony constraints, we are working on that on another PR. I will add the todo in doc.

Comment (greg0ire) [reply FragmentService/AbstractFragmentService.php:124]: ok

**Comment by qsomazzi** [comment]
@greg0ire I've split it in 3 (to be more precise)

File: FragmentService/FragmentServiceInterface.php [hunk 5]
```diff
@@ -42,8 +42,13 @@ public function buildCreateForm(FormMapper $form, FragmentInterface $fragment);
      *
      * @param FragmentInterface         $fragment
      * @param ExecutionContextInterface $context
+    /**
```
Comment (greg0ire) [inline FragmentService/FragmentServiceInterface.php:48]: Uuuh… what?

File: Model/AbstractFragment.php [hunk 6]
```diff
@@ -49,18 +49,23 @@
     protected $position;
 
     /**
-     * @var int
+     * @var string
+     */
+    protected $backofficeTitle;
```
Comment (greg0ire) [inline Model/AbstractFragment.php:54]: Should be private

File: Resources/views/FragmentAdmin/form.html.twig [hunk 7]
```diff
@@ -1,6 +1,6 @@
 {% block form %}
     {% set fragmentName = admin.fragmentServices[form.vars.data.type].name %}
-    <div data-fragment-form="{{ form.vars.id }}" data-form-tmp="true" data-formdata='{ "name": "Fragment {{ fragmentName }}", "type": "{{ fragmentName }}" }' data-errors="{{ form.vars.errors|length }}">
+    <div data-fragment-form="{{ form.vars.id }}" data-form-tmp="true" data-formdata='{ "name": "Fragment {{ fragmentName }}", "type": "{{ form.vars.data.backofficeTitle }}" }' data-errors="{{ form.vars.errors|length }}">
```
Comment (greg0ire) [inline Resources/views/FragmentAdmin/form.html.twig]: Linebreak this please

File: Resources/views/FragmentAdmin/form.html.twig [hunk 8]
```diff
@@ -9,7 +9,7 @@
                     </h4>
                 </div>
                 <div class="box-body">
-                    {% if form.vars.errors|length > 0 %}
+                    {% if form.vars.errors is defined and form.vars.errors|length > 0 %}
```
Comment (greg0ire) [inline Resources/views/FragmentAdmin/form.html.twig]: `form.vars.errors|default([])|lenght` would be shorter, but the best would be to have this variable always defined

File: Resources/doc/reference/todo.rst [hunk 9]
```diff
@@ -0,0 +1,11 @@
+Todo
+====
+
+This bundle is still a work in progress, we need to work more on some part.
```
Comment (greg0ire) [inline Resources/doc/reference/todo.rst]: "parts"

**Review by greg0ire — CHANGES_REQUESTED** [review]

File: Model/AbstractFragment.php (hunk 6 above)
Comment (qsomazzi) [reply Model/AbstractFragment.php:54]: can't set it private, all other properties are protected, and we used them (on fixtures for exemple)

Comment (greg0ire) [reply Model/AbstractFragment.php:54]: ok leave it as is then

File: Resources/views/FragmentAdmin/form.html.twig [hunk 10]
```diff
@@ -9,7 +12,7 @@
                     </h4>
                 </div>
                 <div class="box-body">
-                    {% if form.vars.errors|length > 0 %}
+                    {% if form.vars.errors|default([])|lenght > 0 %}
```
Comment (greg0ire) [inline Resources/views/FragmentAdmin/form.html.twig]: Did you test this? Beware, this is a trick question…

Comment (qsomazzi) [reply Resources/views/FragmentAdmin/form.html.twig]: honestly no, i've trust your knowledge :) will do give me 2min :p

Comment (greg0ire) [reply Resources/views/FragmentAdmin/form.html.twig]: Good answer ;) I mispellt `length`, and you copied it, that's why I'm asking

Comment (qsomazzi) [reply Resources/views/FragmentAdmin/form.html.twig]: @greg0ire all good now :)

**Review by greg0ire — APPROVED** [review]

**Comment by greg0ire** [comment]
@sonata-project/contributors , please review and merge

**Review by rande — APPROVED** [review]

**Comment by OskarStark** [comment]
I mark this PR as `minor` instead of `major`, as there is no current release

**Pull request merged**